text/3550-new-range.md

```diff
@@ -0,0 +1,441 @@
+- use in `Copy` types (currently people work around this by using a tuple instead)
+
+Another primary motivation is the extra size of `RangeInclusive`. It uses an extra `bool` field to keep track of when the upper bound has been yielded by the iterator, but this extra size is useless when the type is not used as an iterator.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Rust has several different types of "range" syntax, including the following:
+
+- `a..b` denotes a range from `a` (inclusive) to `b` (exclusive). It resolves to the type `std::ops::range::Range`.  
+  The iterator for `Range` will yield values from `a` (inclusive) to `b` (exclusive) in steps of one.
+
+- `a..=b` denotes a range from `a` (inclusive) to `b` (inclusive). It resolve to the type `std::ops::range::RangeInclusive`.
+  The iterator for `RangeInclusive` will yield values from `a` (inclusive) to `b` (inclusive) in steps of one.
+
+- `a..` denotes a range from `a` (inclusive) with no upper bound. It resolves to the type `std::ops::range::RangeFrom`.  
+  The iterator for `RangeFrom` will yield values starting with `a` and increasing in steps of one.
+
+These types implement the `IntoIterator` trait, enabling their use directly in a `for` loop:
+```rust
+for n in 0..5 {
+    // `n` = 0, 1, 2, 3, 4
+}
+```
+
+All range types are `Copy` when the bounds are `Copy`, allowing easy reuse:
+```rust
+let range = 0..5;
+
+if a_slice[range].contains(x) {
+    // ...
+}
+if b_slice[range].contains(y) {
+    // ...
+}
+```
+
+For convenience, several commonly-used methods from `Iterator` are present as inherent functions on the range types:
+```rust
+for n in (1..).map(|x| x * 2) {
+    // n = 2, 4, 6, 8, 10, ...
+}
+for n in (0..5).rev() {
+    // n = 4, 3, 2, 1, 0
+}
+```
+
+## Legacy Range Types
+
+In Rust editions prior to 2024, `a..b`, `a..=b`, and `a..` resolved to a different set of types (now found in `std::ops::range::legacy`). These legacy range types did not implement `Copy`, and implemented `Iterator` directly (rather than `IntoIterator`).
+
+This meant that any `Iterator` method could be called on those range types:
+```rust
+let mut range = 0..5;
+assert_eq!(range.next(), Some(0));
+range.for_each(|n| {
+    // n = 1, 2, 3, 4
+});
+```
+
+There exist `From` impls for converting from the new range types to the legacy range types.
+
+### Migrating
+
+`cargo fix --edition 2024` can handle migrating many use cases to the new range types. This includes most places where a `RangeBounds` or `Iterator` is expected:
+```rust
+pub fn takes_range(range: impl std::ops::RangeBounds<usize>) { ... }
+takes_range(0..5); // No changes necessary
+```
+
+And most places where `Iterator` methods were used directly on the range:
+```rust
+for n in (0..5).rev() { ... } // No changes necessary
+
+// Before
+(0..5).for_each(...);
+// After
+(0..5).into_iter().for_each(...); // Add `.into_iter()`
+```
```

Comment:
I think `(0..5).rev()` winds up more common than the pains this RFC addresses, so this change should be avoided somehow,

Reply:
In Guide level explanation, I address this:

> For convenience, several commonly-used methods from Iterator are present as inherent functions on the range types:

```rust
for n in (1..).map(|x| x * 2) {
    // n = 2, 4, 6, 8, 10, ...
}
for n in (0..5).rev() {
    // n = 4, 3, 2, 1, 0
}
```

Reply:
I would favor NOT having this convenience methods, in general.

No other type has such methods provided. You don't type `vec.rev()`, nor `btree_map.rev()`, etc... so it would be a fairly bizarre precedent to have those methods on `Range` and its ilk.

Further, it seems like a potential source of confusion. While `rev` _may_ be relatively clear -- and even then, I could imagine a user thinking it would return a range with swapped bounds -- `map` would be quite unclear. I've myself coded free functions to map the _bounds_, and would definitely read `range.map(...)` as returning an equivalent _range_, not an iterator (and curse whoever made it return an iterator).

If verbosity is a problem, an inherent `iter` method seems a better fit, shaving 5 characters over `into_iter`.

Reply:
I suggest `for x in (0..5).rev()` ets should continue working without changes, including ecosystem extensions like `iterator_ext::IteratorExt`.  We could've some more magic conversion from `0..5` to the iterator form probably.

Reply:
I would argue that `range.rev()` is much more natural than `vec.rev()` considering how ranges would be `Copy` and it's not ambiguous whether there's a mutation here (by convention, `Copy` types don't have self-mutating methods, and if they do, they're clearly labelled as such).

Compare this to `vec.rev()`. It's not clear whether this should be equivalent to `vec.reverse()` or `vec.iter().rev()`. For `BTreeMap` it makes slightly more sense since the collection is always ordered, and it would never make sense to "reverse" the order of a `BTreeMap` in place without changing types.

Reply:
Relevant to this: https://github.com/rust-lang/rfcs/pull/3550/files#r1442993421

> The set of inherent methods copied from Iterator present on the new range types is left open for the libs-api team to decide after this proposal is accepted and before stabilization.

IMHO `range.rev()` is confusing, it's not obvious to me which one of the following should compile:

```rust
for _ in range.rev().skip(1) { ... } 
// or 
for _ in range.rev().into_iter().skip(1) { ... } 
```

This RFC is awesome because it makes `Range` obvious and less special, adding `.rev()` makes it a bit special again, we already have `Iterator::rev` so there's no need for it.

Reply:
> IMHO `range.rev()` is confusing, it's not obvious to me which one of the following should compile

Well, both would compile. `.rev()` returns an `Iterator`, but all iterators implement `IntoIterator`.

That said, I really don't understand how that is confusing. It retains behavior that ranges currently have.

> we already have `Iterator::rev` so there's no need for it.

You won't be able to call `rev` directly, which is a very common case and therefore deserves a convenience method.

Reply:
> I would favor NOT having this convenience methods, in general.

Agree. It's a fantastic RFC but I just find these inherent impls a bit... dirty? My opinion is like "We need to edit the existing code on transition anyway and the fix is rather straightforward. Just less `.into_iter` insertion seems not worthwhile for the weirdness budget."

It can still be decided later though and I'm glad to see one thing I dreamed for finally being pushed!

Reply:
One potential option, if we want to limit breakage but discourage use of these methods, might be to include them but immediately lint (deprecation, FCW?). Then encourage users to instead use `into_iter().foo()`. Though thinking about it, this does have the issue that it's unlikely the methods ever will be removed, since last I knew that can't currently be done in an edition-specific manner.

Reply:
> We need to edit the existing code on transition anyway and the fix is rather straightforward (even can be done automatically by simply cargo fix.) Just less .into_iter insertion seems not worthwhile for the weirdness budget.

It's not just the edition transition we should consider, though. In fact, that's practically irrelevant since it can be fully automated.

We must also consider people writing code by hand. `.into_iter().map()` is 12 extra characters and **14** extra key presses. `map` and `rev` are extremely common and that will add up.

It's also more to read and a significant downgrade from current code. The convenience more than outweighs the weirdness IMO.

Reply:
> It's not just the edition transition we should consider, though.

Make sense. But I'll keep my opinion on "`map` has a better semantic to return a mapped `Range`" ^_^

Reply:
A couple of comments about these convenience methods:

1. AFAIU, the code conversion to insert `.into_iter()` cannot be fully automated. `cargo fix` will not automatically update all the StackOverflow answers doing `(0..n).rev()`, or all the blog posts, the books, the unmaintained OSS projects, etc. All those things will still exist, and will still be read by people that are going to be learning that `(0..n).rev()` is a thing, and then maybe discovering that they were fed lies. This is the kind of language fragmentation that has plagued other languages, and i would not like to see Rust copying.

2. If explicit `range.into_iter()` calls are going to be needed, at least for some cases, could we at least have a convenience `.iter()` method that does the same thing too? 
    Not only for convenience sake, but also for readability and ease of learning. "Going down from N to 0" is a very basic programming task, and it would be a pity if someone just starting with Rust would have to learn about pass-by-move methods just to accomplish the very basic task of printing the numbers from 10 to 1 for example.

Reply:
> 1. All those things will still exist, and will still be read by people that are going to be learning that `(0..n).rev()` is a thing, and then maybe discovering that they were fed lies.

Even with these inherent impls, we can still not deal with all cases. IMO letting beginners find the code they see on the old materials works only occasionally is too surprising and magical, which is not necessarily better than letting them realize the problem at the first place. The latter approach may introduce less confusion and ultimately be better for "how we teach this."